lib/Reader/abisignature.cpp

```diff
@@ -580,7 +581,38 @@ Value *ABICallSignature::emitCall(GenIR &Reader, Value *Target, bool MayThrow,
   Type *FunctionTy = FunctionType::get(FuncResultType, ArgumentTypes, IsVarArg);
   Type *FunctionPtrTy = Reader.getUnmanagedPointerType(FunctionTy);
 
-  Target = Builder.CreateIntToPtr(Target, FunctionPtrTy);
+  // If we're passed a function value, we might need to update the type here
+  // to match, since we didn't know the right type when we created the function.
+  if (isa<llvm::Function>(Target)) {
+    llvm::Function *TargetFunc = cast<llvm::Function>(Target);
+    if (TargetFunc->getType() != FunctionPtrTy) {
+      bool MutateType = false;
+      // If we're going to Jmp to the target, don't mutate the type,
+      // since we may have added parameters to satisfy the constraints
+      // of LLVM's musttail that would cause this signature to diverge
+      // from what we'd have come up with in a normal call to the target.
+      if (!IsJmp) {
+        // We shouldn't change our view of function types multiple times,
+        // so only modify the type if it's the placeholder type we installed
+        // in makeDirectCallTargetNode.
+        Type *VoidType = Type::getVoidTy(Context);
+        PointerType *PlaceholderTy =
+            FunctionType::get(VoidType, false)->getPointerTo();
+        MutateType = TargetFunc->getType() == PlaceholderTy;
```

Comment:
If we see two different signatures for the same function, and the first one happens to be void(), this will wind up mutating the type out from under IR that's already been generated with the previous type in mind.  Do you think that's worth worrying about, or that it can't/won't happen in practice?